src/sentry/backup/sanitize.py

```diff
@@ -0,0 +1,466 @@
+from copy import deepcopy
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone
+from random import choice, randint
+from typing import Callable, List, Optional, Tuple
+
+import petname
+from dateutil.parser import parse as parse_datetime
+from django.utils.text import slugify
+
+from sentry.utils.json import JSONData
+
+UPPER_CASE_HEX = {"A", "B", "C", "D", "E", "F"}
+UPPER_CASE_NON_HEX = {
+    "H",
+    "I",
+    "J",
+    "K",
+    "L",
+    "M",
+    "N",
+    "O",
+    "P",
+    "Q",
+    "R",
+    "S",
+    "T",
+    "U",
+    "V",
+    "W",
+    "X",
+    "Y",
+    "Z",
+}
+LOWER_CASE_HEX = {c.lower() for c in UPPER_CASE_HEX}
+LOWER_CASE_NON_HEX = {c.lower() for c in UPPER_CASE_NON_HEX}
+
+
+class SanitizationError(Exception):
+    """
+    A catch-all class for sanitization errors.
+    """
+
+    pass
+
+
+class InvalidJSONError(SanitizationError):
+    """
+    Thrown when the supplied JSON is not recognizable as a serialized Django model.
+    """
+
+    pass
+
+
+class UnexpectedModelError(SanitizationError):
+    """
+    Thrown when the model does not match that specified by the field supplied by a `set_*`
+    function caller.
+    """
+
+    pass
+
+
+class UnrecognizedDatetimeError(SanitizationError):
+    """
+    Thrown when we encounter a date that has not been interned at startup time.
+    """
+
+    pass
+
+
+@dataclass
+class SanitizableField:
+    """
+    A pairing a of a `NormalizedModelName` with a field in that model, specifying the target for a sanitization operation.
+    """
+
+    from sentry.backup.dependencies import NormalizedModelName
+
+    model: NormalizedModelName
+    field: str
+
+    def validate_json_model(self, json: JSONData) -> None:
+        """
+        Validates the the JSON model is shaped the way we expect a serialized Django model to be,
+        and that we have the right kind of model for this `SanitizableField`. Raises errors if there
+        is a validation failure.
+        """
+
+        model_name = json.get("model", None)
+        if model_name is None:
+            raise InvalidJSONError(
+                "JSON is not properly formatted, must be a serialized Django model"
+            )
+        if model_name != str(self.model):
+            raise UnexpectedModelError(f"Expected model `{model_name}`, got `{str(self.model)}`")
+        return None
+
+
+def _get_field_value(json: JSONData, field: SanitizableField) -> JSONData | None:
+    return json.get("fields", {}).get(field.field, None)
+
+
+def _set_field_value(json: JSONData, field: SanitizableField, value: JSONData) -> JSONData:
+    json.get("fields", {})[field.field] = value
+    return value
+
+
+def default_string_sanitizer(old: str) -> str:
+    """
+    Default string randomizer. Looks at the characters present in the source string to create a new, random string from a roughly similar set of characters.
+    """
+
+    has_upper_case_hex = False
```

Comment:
Doesn't hex include digits too? Guess it doesn't quite matter here since we check for isdigit later on, but a bit confusing to follow along.

Reply:
Yeah, I was hoping the "upper case" portion would make clear that it only refers to letters. It does effect randomization a bit since if we get an `a-f`-only hex string, we'll output an `a-f`-only as well, but I think that's the best we can do from automatic inference alone.